File: src/main/java/discretemath/structure/set/BigDecimalRangeSet.java
package discretemath.structure.set;

import java.math.BigDecimal;
import java.util.Iterator;

import discretemath.structure.exception.UncountableSetException;
import discretemath.structure.set.range.Range;
import discretemath.structure.tuple.Pair;

public class BigDecimalRangeSet extends AbstractSet<BigDecimal> implements RangeSet<BigDecimal> {

	private final Range<BigDecimal> range;

	public BigDecimalRangeSet(Range<BigDecimal> range) {
		this.range = range;
	}

	@Override
	public int size() {
		return -1;
	}

	@Override
	public boolean contains(BigDecimal item) {
		return range.contains(item);
	}

	@Override
	public Set<BigDecimal> unionWith(Set<BigDecimal> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public Set<BigDecimal> intersectWith(Set<BigDecimal> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public Set<BigDecimal> differenceWith(Set<BigDecimal> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public Set<BigDecimal> symmetricDifferenceWith(Set<BigDecimal> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public <P> Set<Pair<BigDecimal, P>> cartesianProductWith(Set<P> set) {
		throw new UnsupportedOperationException("Not implemented yet"); // TODO
	}

	@Override
	public Iterator<BigDecimal> iterator() {
		throw new UncountableSetException();
	}

	@Override
	public Range<BigDecimal> range() {
		return range;
	}
}

File: src/main/scala/discretemath/structure/set/main/BigDecimalRangeSet.scala
package discretemath.structure.set.main

import discretemath.structure.exception.UncountableSetException
import discretemath.structure.set.RangeSet
import discretemath.structure.set.range._

import scala.math.BigDecimal

class BigDecimalRangeSet(val range: Range[BigDecimal]) extends AbstractSet[BigDecimal] with RangeSet[BigDecimal] {

  override def cardinality: Int = -1

  override def contains(item: BigDecimal): Boolean = range.contains(item)

  override def unionWith[I <: Set[BigDecimal], R >: Set[BigDecimal]](set: I): R = ??? // TODO

  override def intersectWith[I <: Set[BigDecimal], R >: Set[BigDecimal]](set: I): R = ???

  override def differenceWith[I <: Set[BigDecimal], R >: Set[BigDecimal]](set: I): R = ???

  override def symmetricDifferenceWith[I <: Set[BigDecimal], R >: Set[BigDecimal]](set: I): R = ???

  override def cartesianProductWith[P](set: Set[P]): Set[(BigDecimal, P)] = ???

  override def iterator = throw new UncountableSetException

  override def powerSet: Set[Set[BigDecimal]] = ???
}
